### PR TITLE
rgw/sfs: Add sqlite schema for sfs metadata

### PR DESCRIPTION
**Pull request by 0xavi0**
Adds initial `sqlite` schema for storing the metadata used by `sfs`.

It only adds the `sqlite` layer, nothing is changed in the `sfs` implementation to not interfere with other possible PRs that are active. (See examples on how to load/store a bucket)

No triggers have beed added to remove objects referencing foreign keys that are removed to not interfere with the `sal` layer, but that is supported in `sqlite_orm`, so we can add it if needed.

Changes in relation to the schema diagram from https://github.com/aquarist-labs/s3gw-core/issues/91:

* Bucket stores all the attributes found in `RGWBucketInfo`, but only a few are really stored. We should store the rest as we need them. I just stored the ones used in `sfs` so far.
* Bucket holds the owner information. This means all objects in a bucket share the owner.
* Bucket ID is the bucket name (Primary key for Buckets table), not an integer as described in the schema diagram.
* `ObjectState` is defined as a C++ enum that is stored as a `unit` in the database. There is a conversion function that also checks that the values in the database are valid enum values.
* `ObjectState` is only present in `VersionedObject`

### Storing a bucket
```c++
#include "rgw/store/sfs/sqlite/sqlite_buckets.h"
using namespace rgw::sal::sfs::sqlite;

auto db_buckets = std::make_shared<SQLiteBuckets>(ceph_context);
DBOPBucketInfo db_bucket_info;
db_bucket_info.binfo = info;  // info is the member in SFSBucket holding the RGWBucketInfo
db_buckets->storeBucket(db_bucket_info);
```

### Loading a bucket
```c++
#include "rgw/store/sfs/sqlite/sqlite_buckets.h"
using namespace rgw::sal::sfs::sqlite;

auto db_buckets = std::make_shared<SQLiteBuckets>(ceph_context);
auto db_bucket_info = db_buckets->getBucket(BUCKET_NAME);
if (db_bucket_info) {
    info = db_bucket_info->binfo;  // store info in SFSBucket 
} else {
    // Bucket was not found, log errors....
}
```

## Code location

All code is located under the `sfs/sqlite` folder expect for the new enum type `ObjectState` that will stored possibly in the `types.h` file from this PR: https://github.com/aquarist-labs/ceph/pull/17

Main classes for accessing the `sqlite` data are located here:

* `sqlite/sqlite_users.h`
* `sqlite/sqlite_buckets.h`
* `sqlite/sqlite_objects.h`
* `sqlite/sqlite_versioned_objects.h`

There is a folder for each table, in which we can find the definition of the objects in the table, the definition of the type that `sfs` will use (normally with ceph types) and the conversions functions for that specific table.
* `sqlite/users/**` 
* `sqlite/buckets/**`
* etc...

The main schema representing the database in C++ is located at:
* `sqlite/sqlite_schema.h` 

Complex `ceph` types are converted using the already existing _encode/decode_ functions from `rgw` or `ceph` base.

Fixes: https://github.com/aquarist-labs/s3gw-core/issues/91

Signed-off-by: Xavi Garcia <xavi.garcia@suse.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/store/sfs/sqlite/versioned_object/versioned_object_conversions.cc [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ */
+#include "versioned_object_conversions.h"
+#include "../conversion_utils.h"
+
+namespace rgw::sal::sfs::sqlite  {
+
+ObjectState getObjectState(uint int_state) {
```
Comment (jecluis) [inline src/rgw/store/sfs/sqlite/versioned_object/versioned_object_conversions.cc:19]: could we rename this `int_state` to something more descriptive? Like `obj_state`?

File: src/test/rgw/sfs/test_rgw_sfs_sqlite_versioned_objects.cc [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+  ASSERT_FALSE(ret_object.has_value());
+}
+
+TEST_F(TestSFSSQLiteVersionedObjects, RemoveObjectThatDoesNotExist) {
+  auto ceph_context = std::make_shared<CephContext>(CEPH_ENTITY_TYPE_CLIENT);
+  ceph_context->_conf.set_val("rgw_sfs_data_path", getTestDir());
+
+  EXPECT_FALSE(fs::exists(getDBFullPath()));
+
+  // Create the object, we need if because of foreign key constrains
```
Comment (jecluis) [inline src/test/rgw/sfs/test_rgw_sfs_sqlite_versioned_objects.cc:247]: s/if/it/ ?

File: src/test/rgw/sfs/test_rgw_sfs_sqlite_versioned_objects.cc [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+
+  // update the size
+  versioned_object.size = 1999;
+  db_versioned_objects->storeVersionedObject(versioned_object);
+
+  ret_ver_object = db_versioned_objects->getVersionedObject(versioned_object.id);
+  ASSERT_TRUE(ret_ver_object.has_value());
+  ASSERT_EQ(1999, ret_ver_object->size);
+  compareVersionedObjects(versioned_object, *ret_ver_object);
+}
```
Comment (jecluis) [inline src/test/rgw/sfs/test_rgw_sfs_sqlite_versioned_objects.cc:298]: shouldn't we be checking that there exists more than one entry for this object?

**Review by jecluis — CHANGES_REQUESTED** [review]
looks good overall, but a few things of note:

* few typos;
* I wonder if the versioned objects test should be checking whether new versions are being created and maintained;
* camel case is being used everywhere while the code base is mostly snake case.

File: src/rgw/store/sfs/sqlite/conversion_utils.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ */
+#pragma once
+
+namespace rgw::sal::sfs::sqlite  {
+
+template <typename BLOB_HOLDER, typename DEST>
+void decodeBlob(const BLOB_HOLDER & blob_holder, DEST & dest) {
```
[the remark this answers is not in the document]
Comment (0xavi0) [reply src/rgw/store/sfs/sqlite/conversion_utils.h:19]: sure, I'll change it.

File: src/rgw/store/sfs/sqlite/objects/object_definitions.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+#include "rgw_common.h"
+
+#include <string>
+
+namespace rgw::sal::sfs::sqlite  {
+
+using BLOB = std::vector<char>;
+
+struct DBObject {
+  std::string ObjectID;
```
[the remark this answers is not in the document]
Comment (0xavi0) [reply src/rgw/store/sfs/sqlite/objects/object_definitions.h:25]: In this case I did it this way to reflect that it is a member representing something in the db table.
To make the member be named like the column in the data base. But no problem on changing it.

File: src/rgw/store/sfs/sqlite/versioned_object/versioned_object_conversions.cc (hunk 1 above)
Comment (0xavi0) [reply src/rgw/store/sfs/sqlite/versioned_object/versioned_object_conversions.cc:19]: sure, I tried to name it int_ to reflect that the input value is a integer value from the db, not an enum, but I'll change it

File: src/test/rgw/sfs/test_rgw_sfs_sqlite_versioned_objects.cc [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+
+TEST_F(TestSFSSQLiteVersionedObjects, CreateAndGet) {
+  auto ceph_context = std::make_shared<CephContext>(CEPH_ENTITY_TYPE_CLIENT);
+  ceph_context->_conf.set_val("rgw_sfs_data_path", getTestDir());
+
+  EXPECT_FALSE(fs::exists(getDBFullPath()));
+  auto db_versioned_objects = std::make_shared<SQLiteVersionedObjects>(ceph_context.get());
+
+  // Create the object, we need if because of foreign key constrains
```
[the remark this answers is not in the document]
Comment (0xavi0) [reply src/test/rgw/sfs/test_rgw_sfs_sqlite_versioned_objects.cc:131]: yup, will change it, thanks!

File: src/rgw/store/sfs/sqlite/objects/object_definitions.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (jecluis) [reply src/rgw/store/sfs/sqlite/objects/object_definitions.h:25]: I'm honestly not very well versed in dealing with databases, so I'm not familiar with what the good customs are for column names, but wouldn't `object_id` be a perfectly valid name as well?

File: src/test/rgw/sfs/test_rgw_sfs_sqlite_versioned_objects.cc (hunk 3 above)
Comment (0xavi0) [reply src/test/rgw/sfs/test_rgw_sfs_sqlite_versioned_objects.cc:307]: I've added that

**Review by jecluis — APPROVED** [review]

**Review by jecluis — APPROVED** [review]

**Pull request merged**